invisible_cities/evm/pmaps.pyx

```diff
@@ -0,0 +1,272 @@
+    The rationale to use s1 and s2 rather than a single
+    class s12 to represent both s1 and s2 is that, although
+    structurally identical, s1 and s2 represent quite
+    different objects. In Particular an s2si is constructed
+    with a s2 not a s1.
+    An s12 is represented as a dictinary of Peaks.
+
+    """
+    def __init__(self, dict s12d):
+
+        cdef int peak_no
+        cdef np.ndarray[double, ndim=1] t
+        cdef np.ndarray[double, ndim=1] E
+        self.peaks = {}
+
+        #print('s12d ={}'.format(s12d))
+        for peak_no, (t, E) in s12d.items():
+            #print('t ={}'.format(t))
+            #print('E ={}'.format(E))
+            assert len(t) == len(E)
+            #p = Peak(t,E)
+            #print('peak = {}'.format(p))
+
+            self.peaks[peak_no] =  Peak(t, E)
+
+    property number_of_peaks:
+         def __get__(self): return len(self.peaks)
+
+    cpdef peak_collection(self):
+        try:
+            return tuple(self.peaks.keys())
+        except KeyError:
+            raise PeakNotFound
+
+    cpdef peak_waveform(self, int peak_number):
+        try:
+            return self.peaks[peak_number]
+        except KeyError:
+             raise PeakNotFound
+
+    cpdef store(self, table, event_number):
+        row = table.row
+        for peak_number, peak in self.peaks.items():
+            for t, E in zip(peak.t, peak.E):
+                row["event"] = event_number
+                row["peak"]  =  peak_number
+                row["time"]  = t
+                row["ene"]   = E
+                row.append()
+
+
+cdef class S1(S12):
+    def __init__(self, s1d):
+        self.s1d = s1d
+        super(S1,self).__init__(s1d)
+
+    def __str__(self):
+        s =  "S1 (number of peaks = {})\n".format(self.number_of_peaks)
+        s2 = ['peak number = {}: {} \n'.format(i,
+                                    self.peak_waveform(i)) for i in self.peaks]
+        return  s + ''.join(s2)
+
+    def __repr__(self):
+        return self.__str__()
+
+cdef class S2(S12):
+    def __init__(self, s2d):
+        self.s2d = s2d
+        super(S2, self).__init__(s2d)
+
+    def __str__(self):
+        s =  "S2 (number of peaks = {})\n".format(self.number_of_peaks)
+        s2 = ['peak number = {}: {} \n'.format(i,
+                                    self.peak_waveform(i)) for i in self.peaks]
+        return  s + ''.join(s2)
+
+    def __repr__(self):
+        return self.__str__()
+
+cdef class S2Si(S2):
+    """Transient class representing the combination of
+    S2 and the SiPM information.
+    Notice that S2Si is constructed using an s2d and an s2sid.
+    The s2d is an s12 dictionary (not an S2 instance)
+    The s2sid is a dictionary {peak:{nsipm:[E]}}
+    """
+
+    def __init__(self, s2d, s2sid):
+        """where:
+           s2d   = {peak_number:[[t], [E]]}
+           s2sid = {peak:{nsipm:[Q]}}
+           Q is the energy in each SiPM sample
+        """
+        S2.__init__(self, s2d)
```

Comment:
It is redundant but nice. Consider the case in which you want to retrieve the waveform of a given SiPM. The underlying s2sid object has only information on E but not on time (since this is own by the s2d object). By keeping a reference to the s2d object one can access the time waveform (and also the energy waveform) of the s2 *directly*, and this is provided by the high level interface without the need for gimmnastics (e.g, go and search for the corresponding information in the s2 object). At the same time, by keeping the reference to s2, we do not need to save the time waveform for each SiPM in the S2Si, thus saving space and speed access.